hostlists-builder/index.js

```diff
@@ -193,42 +133,63 @@ const loadLocales = function (dir) {
   const filtersMetadata = [];
   const filtersMetadataDev = [];
+  const filterKeyValidator = filterKeyValidatorFactory();
+  const deferredRunner = new DeferredRunner();
+  const tagsMetadata = JSON.parse(await readFile(path.join(tagsDir, METADATA_FILE)));
+  const tagsMetadataUtils = new TagsMetadataUtils(tagsMetadata.slice());
 
-  const filterDirs = listFiltersDirs(filtersDir);
+  const filterDirs = await listFiltersDirs(filtersDir);
   for (const filterDir of filterDirs) {
-    const metadata = JSON.parse(readFile(path.join(filterDir, METADATA_FILE)));
+    const metadata = JSON.parse(await readFile(path.join(filterDir, METADATA_FILE)));
+
+    // Validate filterKey field
+    filterKeyValidator.validate(metadata.filterKey);
 
     // Reads the current revision information.
     const revisionFile = path.join(filterDir, REVISION_FILE);
-    const currentRevision = JSON.parse(readFile(revisionFile)) || { timeUpdated: new Date().getTime() };
-    let timeUpdated = currentRevision.timeUpdated;
+    const revision = new Revision(JSON.parse(await readFile(revisionFile)));
+
+    if (typeof metadata.filterId === 'undefined') {
+      throw new Error('You should use `filterId` instead of `id` in metadata.json');
+    }
 
     // Compiles the hostlist using provided configuration.
-    const hostlistConfiguration = readHostlistConfiguration(filterDir);
-    const filterName = `filter_${metadata.id}.txt`;
+    const hostlistConfiguration = await readHostlistConfiguration(filterDir);
+    const filterName = `filter_${metadata.filterId}.txt`;
 
     // If the hostlist is disabled, do not attempt to download it, just use the
     // existing one.
     if (!metadata.disabled) {
       try {
-        const hostlistCompiled = await hostlistCompiler(hostlistConfiguration);
+        revision.safelyIncrementVersion();
+        revision.safelyUpdateTime();
+
+        const hostlistCompiled = await hostlistCompiler({
+          ...hostlistConfiguration,
+          version: revision.version,
+        });
+
         const hash = calculateRevisionHash(hostlistCompiled);
 
         // Rewrites the filter if it's actually changed.
-
-        if (currentRevision.hash !== hash) {
-          const newRevision = makeRevision(currentRevision, hash);
+        if (revision.hash !== hash) {
+          revision.setHash(hash);
           const assetsFilterFile = path.join(assetsDir, filterName);
           const filterFile = path.join(filterDir, 'filter.txt');
-          let content = hostlistCompiled.join('\n');
-
-          timeUpdated = newRevision.timeUpdated;
-          writeFile(revisionFile, JSON.stringify(newRevision, null, '\t'));
-          writeFile(assetsFilterFile, content);
-          writeFile(filterFile, content);
+          const content = hostlistCompiled.join('\n');
+
+          // We don't write files now, cause next iteration may fails.
+          // We want do it clear after loop
+          deferredRunner.push(async () => {
```

Comment:
Please note that now files are written only after all filters have been successfully processed